tests/labels_test.go

```diff
@@ -0,0 +1,130 @@
+package tests
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"kubevirt.io/ssp-operator/internal/common"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"gomodules.xyz/jsonpatch/v2"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func expectedLabelsFor(name string, component common.AppComponent) map[string]string {
+	return map[string]string{
+		common.AppKubernetesNameLabel:      name,
+		common.AppKubernetesManagedByLabel: "ssp-operator",
+		common.AppKubernetesPartOfLabel:    strategy.GetPartOfLabel(),
+		common.AppKubernetesVersionLabel:   strategy.GetVersionLabel(),
+		common.AppKubernetesComponentLabel: component.String(),
+	}
+}
+
+func expectAppLabels(res *testResource) {
+	testExpectedLabels(res.NewResource(), res.GetKey(), res.ExpectedLabels)
+}
+
+func expectAppLabelsRestoreAfterUpdate(res *testResource) {
+	testExpectedLabelsRestoreAfterUpdate(res.NewResource(), res.GetKey(), res.ExpectedLabels)
+}
+
+func testExpectedLabels(resource controllerutil.Object, key client.ObjectKey, expectedLabels map[string]string) {
+	patchAppLabelsIntoSSP(expectedLabels)
+	waitForLabelMatch(resource, key, expectedLabels)
+}
+
+func testExpectedLabelsRestoreAfterUpdate(resource controllerutil.Object, key client.ObjectKey, expectedLabels map[string]string) {
+	testExpectedLabels(resource, key, expectedLabels)
+
+	operations := newLabelOperations(expectedLabels)
+	for label := range expectedLabels {
+		operations = append(operations, labelPatchOperationFor("replace", label, "wrong"))
+	}
+	patch := encodePatch(operations)
+	err := apiClient.Patch(ctx, resource, patch)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(labelsMatch(expectedLabels, resource)).To(BeFalse())
+}
+
+func patchAppLabelsIntoSSP(expectedLabels map[string]string) {
```

Comment:
This function is no longer needed. We can assume the SSP CR contains the labels, as we discussed in one of the other comments.

Reply:
After debugging, I found out why the tests fail when this function is removed.
The reason is that the `createOrUpdateSsp()` function in `tests_suite_test.go` does not recreate the SSP with labels.
Can you change that function to the following?
```
func createOrUpdateSsp(ssp *sspv1beta1.SSP) {
	key := client.ObjectKey{
		Name:      ssp.Name,
		Namespace: ssp.Namespace,
	}
	Eventually(func() error {
		foundSsp := &sspv1beta1.SSP{}
		err := apiClient.Get(ctx, key, foundSsp)
		if err == nil {
			isSspEqual := reflect.DeepEqual(foundSsp.Spec, ssp.Spec) &&
				reflect.DeepEqual(foundSsp.Annotations, ssp.Annotations) &&
				reflect.DeepEqual(foundSsp.Labels, ssp.Labels)
			if isSspEqual {
				return nil
			}
			foundSsp.Spec = ssp.Spec
			foundSsp.Annotations = ssp.Annotations
			foundSsp.Labels = ssp.Labels
			return apiClient.Update(ctx, foundSsp)
		}
		if errors.IsNotFound(err) {
			newSsp := &sspv1beta1.SSP{
				ObjectMeta: metav1.ObjectMeta{
					Name:        ssp.Name,
					Namespace:   ssp.Namespace,
					Annotations: ssp.Annotations,
					Labels:      ssp.Labels,
				},
				Spec: ssp.Spec,
			}
			return apiClient.Create(ctx, newSsp)
		}
		return err
	}, timeout, time.Second).ShouldNot(HaveOccurred())
}
```

And then, you can remove this one `patchAppLabelsIntoSSP`.

Reply:
Thanks for spotting this, I updated the PR.

Reply:
Please also update the `reflect.DeepEqual()` check on line 419 in `tests_suite_test.go`. See my previous comment, the deep equals need to compare labels and annotations, not only spec.

Reply:
Ja, me blind... Thanks again.